parser/options.go

```diff
@@ -0,0 +1,54 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package parser
+
+type options struct {
+	maxRecursionDepth  int
+	errorRecoveryLimit int
+	macros             map[string]Macro
+}
+
+// Option configures the behavior of the parser.
+type Option func(*options)
+
+// MaxRecursionDepth is an option which limits the maximum depth the parser will attempt to
+// parse the expression before giving up.
+func MaxRecursionDepth(maxRecursionDepth int) Option {
+	return func(opts *options) {
+		opts.maxRecursionDepth = maxRecursionDepth
+	}
+}
+
+// ErrorRecoveryLimit is an option which limits the number of attempts the parser will perform
+// to recover from an error.
+func ErrorRecoveryLimit(errorRecoveryLimit int) Option {
+	return func(opts *options) {
+		opts.errorRecoveryLimit = errorRecoveryLimit
+	}
+}
+
+// Macros adds the given macros to the parser.
+func Macros(macros ...Macro) Option {
+	return func(opts *options) {
+		for _, m := range macros {
+			if m != nil {
+				if opts.macros == nil {
+					opts.macros = make(map[string]Macro)
+				}
+				opts.macros[m.MacroKey()] = m
```

Comment:
Consider erroring if the `MacroKey` is used more than once.

Reply:
I avoided doing so in case a caller wanted to replace a builtin macro with their own, they could do something like `Macros(AllMacros...), Macros(MyHasMacro)`.

Reply:
I'm going to hoist the individual macros out of the `AllMacros` set and export them to the `cel` package, so it should be much easier to compose replacements.